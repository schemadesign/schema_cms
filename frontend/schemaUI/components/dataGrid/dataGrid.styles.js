import { defaultTheme } from '../../utils/theme';

export const BLUE_COLOR = '#778DA9';
export const BLACK_COLOR = '#3C474B';
export const WHITE_COLOR = '#fff';
export const GRAY_COLOR = '#E0E1DD';

export const getStyles = (theme = defaultTheme) => ({
  containerStyles: {
    display: 'flex',
    flexDirection: 'column',
    width: '100%',
  },
  columnStyles: {
    columnItem: {
      backgroundColor: theme.dataGrid.background,
      color: theme.dataGrid.text,
      border: '1px solid #fff',
      padding: '10px 15px',
    },
    columnHeader: {
      display: 'flex',
    },
  },
  rowStyles: {
    rowItem: {
      padding: '10px 15px',
    },
    rowHeader: {
      width: '100%',
      display: 'flex',
      flexWrap: 'wrap',
    },
  },
});
