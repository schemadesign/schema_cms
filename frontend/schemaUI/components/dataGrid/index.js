export { DataGrid } from './dataGrid.component';
