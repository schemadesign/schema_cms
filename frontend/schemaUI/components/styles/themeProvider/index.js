export { ThemeProvider } from './themeProvider.component';
