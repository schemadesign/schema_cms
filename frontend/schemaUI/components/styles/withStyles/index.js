export { withStyles } from './withStyles.component';
