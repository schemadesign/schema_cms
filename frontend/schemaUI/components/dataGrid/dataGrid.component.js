import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import { containerStyles, columnStyles, rowStyles } from './dataGrid.styles';
import { Grid } from '../grid';

const SCREEN_WIDTH_PERCENT = '100';

export class DataGrid extends PureComponent {
  static propTypes = {
    data: PropTypes.object.isRequired,
  };

  cellWidths = [];

  filterSizedItems = item => item.size;

  reduceSizeSum = (total, currentColumn) => total + (currentColumn.size || 0);

  getCellWidth = size => {
    const sumSize = this.props.data.columns.reduce(this.reduceSizeSum, 0);
    const sizedItemsCount = this.props.data.columns.filter(this.filterSizedItems).length;

    return size ? size : (SCREEN_WIDTH_PERCENT - sumSize) / (this.props.data.columns.length - sizedItemsCount);
  };

  calculateCellWidth = ({ name, size }) => {
    this.cellWidths[name] = this.getCellWidth(size);
  };

  getCustomStyles = (styles, defaultStyle = {}) => styles || defaultStyle;

  renderColumns = (column, index) => {
    return (
      <Grid
        key={index}
        size={this.cellWidths[column.name]}
        customStyles={this.getCustomStyles(column.styles, columnStyles.columnItem)}
      >
        {column.displayName}
      </Grid>
    );
  };

  renderRows = (row, index) => (
    <Grid key={index} size={100}>
      {this.props.data.columns.map(({ name }, index) => {
        const styles = row.styles ? row.styles[name] : {};

        return (
          <Grid key={index} size={this.cellWidths[name]} customStyles={this.getCustomStyles(styles)}>
            {row[name]}
          </Grid>
        );
      })}
    </Grid>
  );

  render() {
    const {
      data: { columns, rows },
    } = this.props;

    columns.forEach(this.calculateCellWidth);

    return (
      <div style={containerStyles}>
        <div style={columnStyles.columnHeader}>{columns.map(this.renderColumns)}</div>
        <div style={rowStyles.rowHeader}>{rows.map(this.renderRows)}</div>
      </div>
    );
  }
}
