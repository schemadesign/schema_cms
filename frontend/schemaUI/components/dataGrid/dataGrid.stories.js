import React from 'react';
import { storiesOf } from '@storybook/react';

import { DataGrid } from './dataGrid.component';
import { HEADER_COLOR, TEXT_COLOR } from './dataGrid.styles';

const BLACK_COLOR = '#000';

const firstColumnStyle = {
  backgroundColor: HEADER_COLOR,
  color: TEXT_COLOR,
};

const columnStyles = {
  backgroundColor: HEADER_COLOR,
  color: TEXT_COLOR,
};

const rowStyles = {
  id: {
    styles: {
      backgroundColor: HEADER_COLOR,
      color: TEXT_COLOR,
    },
  },
  name: {
    backgroundColor: BLACK_COLOR,
    color: TEXT_COLOR,
  },
  surname: {
    backgroundColor: BLACK_COLOR,
    color: TEXT_COLOR,
  },
  dateOfBirth: {
    backgroundColor: BLACK_COLOR,
    color: TEXT_COLOR,
  },
};

const data = {
  columns: [
    { name: 'id', displayName: '#', styles: firstColumnStyle, size: 10 },
    { name: 'name', displayName: 'Name', styles: columnStyles },
    { name: 'surname', displayName: 'Surname', styles: columnStyles },
    { name: 'dateOfBirth', displayName: 'Date of Birth', styles: columnStyles },
  ],
  rows: [
    { id: '342', name: 'Alan', surname: 'Watts', dateOfBirth: '06/01/1915', styles: rowStyles },
    { id: '355', name: 'David', surname: 'Bowie', dateOfBirth: '08/01/1947', styles: rowStyles },
    { id: '123', name: 'Dale', surname: 'Chihuly', dateOfBirth: '20/09/1941', styles: rowStyles },
    { id: '556', name: 'SebastiÃ£o', surname: 'Salgado', dateOfBirth: '08/02/1944', styles: rowStyles },
  ],
};

const defaultProps = {
  data,
};

storiesOf('DataGrid', module).add('Default', () => <DataGrid {...defaultProps} />);
