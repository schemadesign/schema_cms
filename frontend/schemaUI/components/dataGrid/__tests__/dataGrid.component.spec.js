import React from 'react';
import { mount } from 'enzyme';

import { DataGrid } from '../dataGrid.component';
import { BLACK_COLOR, BLUE_COLOR, WHITE_COLOR, GRAY_COLOR } from '../dataGrid.styles';

describe('DataGrid: Component', () => {
  const data = {
    columns: [
      { name: 'id', displayName: '#' },
      { name: 'name', displayName: 'Name' },
      { name: 'surname', displayName: 'Surname' },
      { name: 'dateOfBirth', displayName: 'Date of Birth' },
    ],
    rows: [
      { id: '342', name: 'Alan', surname: 'Watts', dateOfBirth: '06/01/1915' },
      { id: '355', name: 'David', surname: 'Bowie', dateOfBirth: '08/01/1947' },
      { id: '123', name: 'Dale', surname: 'Chihuly', dateOfBirth: '20/09/1941' },
      { id: '556', name: 'Sebastião', surname: 'Salgado', dateOfBirth: '08/02/1944' },
    ],
  };

  const defaultProps = {
    data,
  };

  const component = props => <DataGrid {...defaultProps} {...props} />;

  const render = (props = {}) => mount(component(props));

  it('should render correctly', () => {
    const wrapper = render();
    global.expect(wrapper).toMatchSnapshot();
  });

  it('should render with custom styles and size', () => {
    const columnAlternativeStyles = {
      backgroundColor: BLUE_COLOR,
      border: `0.3px solid ${WHITE_COLOR}`,
      padding: '10px 5px',
      color: WHITE_COLOR,
    };

    const rowAlternativeStyles = {
      id: {
        backgroundColor: BLUE_COLOR,
        border: `0.3px solid ${WHITE_COLOR}`,
        padding: '10px 5px',
        color: WHITE_COLOR,
      },
      name: {
        backgroundColor: BLACK_COLOR,
        border: `0.3px solid ${WHITE_COLOR}`,
        padding: '10px 5px',
        color: GRAY_COLOR,
      },
      surname: {
        backgroundColor: BLACK_COLOR,
        border: `0.3px solid ${WHITE_COLOR}`,
        padding: '10px 5px',
        color: GRAY_COLOR,
      },
      dateOfBirth: {
        backgroundColor: BLACK_COLOR,
        border: `0.3px solid ${WHITE_COLOR}`,
        padding: '10px 5px',
        color: GRAY_COLOR,
      },
    };

    const dataWithStylesAndSize = {
      columns: [
        { name: 'id', displayName: '# - Sized', styles: columnAlternativeStyles, size: 10 },
        { name: 'name', displayName: 'Name', styles: columnAlternativeStyles },
        { name: 'surname', displayName: 'Surname', styles: columnAlternativeStyles },
        { name: 'dateOfBirth', displayName: 'Date of Birth', styles: columnAlternativeStyles },
      ],
      rows: [
        { id: '342', name: 'Alan', surname: 'Watts', dateOfBirth: '06/01/1915', styles: rowAlternativeStyles },
        { id: '355', name: 'David', surname: 'Bowie', dateOfBirth: '08/01/1947', styles: rowAlternativeStyles },
        { id: '123', name: 'Dale', surname: 'Chihuly', dateOfBirth: '20/09/1941', styles: rowAlternativeStyles },
        { id: '556', name: 'Sebastião', surname: 'Salgado', dateOfBirth: '08/02/1944', styles: rowAlternativeStyles },
      ],
    };

    const wrapper = render({ data: dataWithStylesAndSize });
    global.expect(wrapper).toMatchSnapshot();
  });
});
