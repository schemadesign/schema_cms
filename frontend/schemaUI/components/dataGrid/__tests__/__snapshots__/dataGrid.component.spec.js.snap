// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataGrid: Component should render correctly 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "width": "100%",
    }
  }
>
  <div
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <Grid
      customStyles={
        Object {
          "backgroundColor": "#1D1D20",
          "border": "1px solid #fff",
          "color": "#71737e",
          "padding": "10px 15px",
        }
      }
      key="0"
      size={25}
    >
      #
    </Grid>
    <Grid
      customStyles={
        Object {
          "backgroundColor": "#1D1D20",
          "border": "1px solid #fff",
          "color": "#71737e",
          "padding": "10px 15px",
        }
      }
      key="1"
      size={25}
    >
      Name
    </Grid>
    <Grid
      customStyles={
        Object {
          "backgroundColor": "#1D1D20",
          "border": "1px solid #fff",
          "color": "#71737e",
          "padding": "10px 15px",
        }
      }
      key="2"
      size={25}
    >
      Surname
    </Grid>
    <Grid
      customStyles={
        Object {
          "backgroundColor": "#1D1D20",
          "border": "1px solid #fff",
          "color": "#71737e",
          "padding": "10px 15px",
        }
      }
      key="3"
      size={25}
    >
      Date of Birth
    </Grid>
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "flexWrap": "wrap",
        "width": "100%",
      }
    }
  >
    <Grid
      key="0"
      size={100}
    >
      <Grid
        customStyles={Object {}}
        key="0"
        size={25}
      >
        342
      </Grid>
      <Grid
        customStyles={Object {}}
        key="1"
        size={25}
      >
        Alan
      </Grid>
      <Grid
        customStyles={Object {}}
        key="2"
        size={25}
      >
        Watts
      </Grid>
      <Grid
        customStyles={Object {}}
        key="3"
        size={25}
      >
        06/01/1915
      </Grid>
    </Grid>
    <Grid
      key="1"
      size={100}
    >
      <Grid
        customStyles={Object {}}
        key="0"
        size={25}
      >
        355
      </Grid>
      <Grid
        customStyles={Object {}}
        key="1"
        size={25}
      >
        David
      </Grid>
      <Grid
        customStyles={Object {}}
        key="2"
        size={25}
      >
        Bowie
      </Grid>
      <Grid
        customStyles={Object {}}
        key="3"
        size={25}
      >
        08/01/1947
      </Grid>
    </Grid>
    <Grid
      key="2"
      size={100}
    >
      <Grid
        customStyles={Object {}}
        key="0"
        size={25}
      >
        123
      </Grid>
      <Grid
        customStyles={Object {}}
        key="1"
        size={25}
      >
        Dale
      </Grid>
      <Grid
        customStyles={Object {}}
        key="2"
        size={25}
      >
        Chihuly
      </Grid>
      <Grid
        customStyles={Object {}}
        key="3"
        size={25}
      >
        20/09/1941
      </Grid>
    </Grid>
    <Grid
      key="3"
      size={100}
    >
      <Grid
        customStyles={Object {}}
        key="0"
        size={25}
      >
        556
      </Grid>
      <Grid
        customStyles={Object {}}
        key="1"
        size={25}
      >
        Sebastião
      </Grid>
      <Grid
        customStyles={Object {}}
        key="2"
        size={25}
      >
        Salgado
      </Grid>
      <Grid
        customStyles={Object {}}
        key="3"
        size={25}
      >
        08/02/1944
      </Grid>
    </Grid>
  </div>
</div>
`;

exports[`DataGrid: Component should render with custom styles and size 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "width": "100%",
    }
  }
>
  <div
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <Grid
      customStyles={
        Object {
          "backgroundColor": "#778DA9",
          "border": "0.3px solid #fff",
          "color": "#fff",
          "padding": "10px 5px",
        }
      }
      key="0"
      size={10}
    >
      # - Sized
    </Grid>
    <Grid
      customStyles={
        Object {
          "backgroundColor": "#778DA9",
          "border": "0.3px solid #fff",
          "color": "#fff",
          "padding": "10px 5px",
        }
      }
      key="1"
      size={30}
    >
      Name
    </Grid>
    <Grid
      customStyles={
        Object {
          "backgroundColor": "#778DA9",
          "border": "0.3px solid #fff",
          "color": "#fff",
          "padding": "10px 5px",
        }
      }
      key="2"
      size={30}
    >
      Surname
    </Grid>
    <Grid
      customStyles={
        Object {
          "backgroundColor": "#778DA9",
          "border": "0.3px solid #fff",
          "color": "#fff",
          "padding": "10px 5px",
        }
      }
      key="3"
      size={30}
    >
      Date of Birth
    </Grid>
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "flexWrap": "wrap",
        "width": "100%",
      }
    }
  >
    <Grid
      key="0"
      size={100}
    >
      <Grid
        customStyles={
          Object {
            "backgroundColor": "#778DA9",
            "border": "0.3px solid #fff",
            "color": "#fff",
            "padding": "10px 5px",
          }
        }
        key="0"
        size={10}
      >
        342
      </Grid>
      <Grid
        customStyles={
          Object {
            "backgroundColor": "#3C474B",
            "border": "0.3px solid #fff",
            "color": "#E0E1DD",
            "padding": "10px 5px",
          }
        }
        key="1"
        size={30}
      >
        Alan
      </Grid>
      <Grid
        customStyles={
          Object {
            "backgroundColor": "#3C474B",
            "border": "0.3px solid #fff",
            "color": "#E0E1DD",
            "padding": "10px 5px",
          }
        }
        key="2"
        size={30}
      >
        Watts
      </Grid>
      <Grid
        customStyles={
          Object {
            "backgroundColor": "#3C474B",
            "border": "0.3px solid #fff",
            "color": "#E0E1DD",
            "padding": "10px 5px",
          }
        }
        key="3"
        size={30}
      >
        06/01/1915
      </Grid>
    </Grid>
    <Grid
      key="1"
      size={100}
    >
      <Grid
        customStyles={
          Object {
            "backgroundColor": "#778DA9",
            "border": "0.3px solid #fff",
            "color": "#fff",
            "padding": "10px 5px",
          }
        }
        key="0"
        size={10}
      >
        355
      </Grid>
      <Grid
        customStyles={
          Object {
            "backgroundColor": "#3C474B",
            "border": "0.3px solid #fff",
            "color": "#E0E1DD",
            "padding": "10px 5px",
          }
        }
        key="1"
        size={30}
      >
        David
      </Grid>
      <Grid
        customStyles={
          Object {
            "backgroundColor": "#3C474B",
            "border": "0.3px solid #fff",
            "color": "#E0E1DD",
            "padding": "10px 5px",
          }
        }
        key="2"
        size={30}
      >
        Bowie
      </Grid>
      <Grid
        customStyles={
          Object {
            "backgroundColor": "#3C474B",
            "border": "0.3px solid #fff",
            "color": "#E0E1DD",
            "padding": "10px 5px",
          }
        }
        key="3"
        size={30}
      >
        08/01/1947
      </Grid>
    </Grid>
    <Grid
      key="2"
      size={100}
    >
      <Grid
        customStyles={
          Object {
            "backgroundColor": "#778DA9",
            "border": "0.3px solid #fff",
            "color": "#fff",
            "padding": "10px 5px",
          }
        }
        key="0"
        size={10}
      >
        123
      </Grid>
      <Grid
        customStyles={
          Object {
            "backgroundColor": "#3C474B",
            "border": "0.3px solid #fff",
            "color": "#E0E1DD",
            "padding": "10px 5px",
          }
        }
        key="1"
        size={30}
      >
        Dale
      </Grid>
      <Grid
        customStyles={
          Object {
            "backgroundColor": "#3C474B",
            "border": "0.3px solid #fff",
            "color": "#E0E1DD",
            "padding": "10px 5px",
          }
        }
        key="2"
        size={30}
      >
        Chihuly
      </Grid>
      <Grid
        customStyles={
          Object {
            "backgroundColor": "#3C474B",
            "border": "0.3px solid #fff",
            "color": "#E0E1DD",
            "padding": "10px 5px",
          }
        }
        key="3"
        size={30}
      >
        20/09/1941
      </Grid>
    </Grid>
    <Grid
      key="3"
      size={100}
    >
      <Grid
        customStyles={
          Object {
            "backgroundColor": "#778DA9",
            "border": "0.3px solid #fff",
            "color": "#fff",
            "padding": "10px 5px",
          }
        }
        key="0"
        size={10}
      >
        556
      </Grid>
      <Grid
        customStyles={
          Object {
            "backgroundColor": "#3C474B",
            "border": "0.3px solid #fff",
            "color": "#E0E1DD",
            "padding": "10px 5px",
          }
        }
        key="1"
        size={30}
      >
        Sebastião
      </Grid>
      <Grid
        customStyles={
          Object {
            "backgroundColor": "#3C474B",
            "border": "0.3px solid #fff",
            "color": "#E0E1DD",
            "padding": "10px 5px",
          }
        }
        key="2"
        size={30}
      >
        Salgado
      </Grid>
      <Grid
        customStyles={
          Object {
            "backgroundColor": "#3C474B",
            "border": "0.3px solid #fff",
            "color": "#E0E1DD",
            "padding": "10px 5px",
          }
        }
        key="3"
        size={30}
      >
        08/02/1944
      </Grid>
    </Grid>
  </div>
</div>
`;
